[BugFix] Fixes #20983 - Store py_func dtypes for correct conversion

sabhiram opened the pull request:
1.  Implement wrapper class `PythonFunc` to store Tout list
    for appropriate dtype conversion from tensorflow types to
    numpy types using the `dtype.as_numpy_dtype` member.
2.  Add py_func_test to illustrate broken behavior (see
    `testConvertEmptyList`).

sabhiram committed:
[BugFix] Fixes #20983 - Store py_func dtypes for correct conversion

1.  Implement wrapper class `PythonFunc` to store Tout list
    for appropriate dtype conversion from tensorflow types to
    numpy types using the `dtype.as_numpy_dtype` member.
2.  Adjust BUILD file to include py_test of the script_ops_test.
3.  Add script_ops_test to illustrate broken behavior (see
    `test_convert_empty_list`). Run the specific test suite with:

      bazel test //tensorflow/python:script_ops_test

4.  `_convert` was being called from dataset_ops, update this to
    refer to the `PythonFunc` instead of the `FuncRegistry`.
5.  [Review Feedback] Move script_ops tests to py_func_test.py. Also
    removed the BUILD rules for the deleted test file.

sabhiram commented:
^ WIP on the CI errors - they were due to the `FuncRegistry._convert()` being used from `dataset_ops`.  If this is being called this way, should convert be exposed from `script_ops` more officially (not through the FuncRegistry)?

My fixes are done, I am in-tow for the 9 million bazel files that needed recompilation :) Will squash and push once I confirm things are resolved locally.

akshayka commented on tensorflow/python/ops/script_ops_test.py:
```diff
@@ -0,0 +1,47 @@
+  """Tests for script_ops.py_func."""
+
+  def test_basic_convert(self):
+    def get_not_empty_list(args):
```
Can you delete this file and add your tests to tensorflow/python/kernel_tests/py_func_test.py?

akshayka commented on tensorflow/python/BUILD:
```diff
@@ -2524,6 +2524,18 @@ py_library(
     ],
 )
 
+py_test(
```
Delete this build rule and instead add your tests to tensorflow/python/kernel_tests/py_func_test.py

sabhiram replied:
Can do - thanks for the pointers. Will fix and push shortly.

sabhiram commented:
@akshayka - Thanks for the feedback, I think I should have done justice to your comments. Please let me know if this looks better. 

PS: One of the basic tests that I had originally added seemed useless given the py_func_test.py file, so I went ahead and removed it (preserved the test that will *fail* without these changes).

akshayka commented on tensorflow/python/ops/script_ops.py:
```diff
@@ -171,8 +160,48 @@ def _convert(value, dtype=None):
+          self._convert(x, dtype=dtype.as_numpy_dtype)
+          for (x, dtype) in zip(ret, self._out_dtypes)
+      ]
+    return self._convert(ret)  # TODO: pass self._out_dtypes?
```
Is there any reason to not pass out_dtypes here?

sabhiram replied:
One thing - the registry will force the Tout to be an `array / list` -- see ~line 270 in the `_internal_py_func` where the `Tout` is initialized.  So we would need to sanitize the _out_dtypes. But that should be easy.

sabhiram committed:
[Review Feedback] apply dtype to _convert on non list inputs to py_func

Disabled the `testBadNumpyReturnType` which should allow for a
conversion of an arbitrary np.array into a list of a specified dtype.

Modified the exception expectation for the `testBadReturnType` test.

akshayka commented:
Great, go ahead and delete the test you commented out, I agree with you that it no longer applies.

sabhiram committed:
[Review Feedback] remove commented test

sabhiram commented:
@akshaym - any reservations?

akshaym commented:
@sabhiram not from me. I think @akshayka is the best person to review this, so I'll unassign myself.

sabhiram commented:
@akshayka - I think this is good to go - might need your blessing before the bot takes over.

sabhiram commented:
@akshayka - Heh, what did I break? Unfortunately I cannot peer into the CI process enough to guess how this affected what.  Happy to help see it through if needed.

akshayka commented:
Hey Sabhiram, sorry for the delay. I didn't have the time to diagnose the issue, but the change caused an internal-only test to segfault.
